lib/task.ml

```diff
@@ -1,33 +1,108 @@
-type 'a task = unit -> 'a
+open EffectHandlers
+open EffectHandlers.Deep
 
-type 'a promise = ('a, exn) result option Atomic.t
+type 'a task = unit -> 'a
 
-exception TasksActive
+type message =
+  Work of (unit -> unit)
+| Quit
 
-type task_msg =
-  Task : 'a task * 'a promise -> task_msg
-| Quit : task_msg
+type task_chan = message Multi_channel.t
 
 type pool_data = {
   domains : unit Domain.t array;
-  task_chan : task_msg Multi_channel.t;
+  task_chan : task_chan;
   name: string option
 }
 
 type pool = pool_data option Atomic.t
 
+type 'a promise_state =
+  Returned of 'a
+| Raised of exn * Printexc.raw_backtrace
+| Pending of ((unit, unit) continuation * task_chan) list
+
+type 'a promise = 'a promise_state Atomic.t
+
+type _ eff += Wait : 'a promise * task_chan -> unit eff
+
+let get_pool_data p =
+  match Atomic.get p with
```

Comment:
I think there might be multiple semantics to pool teardown (might be useful to have a function that just waits/drains the active tasks without causing exceptions to be raised), but once you reach teardown it could just go through the queue of all active tasks and "cancel" them by setting their result to be an exception of cancelled (obviously can't cancel any that are actively running, the await might have some critical "finally" statements to clean up resources). But it might allow removing this Atomic.get if some kind of cancelation is implemented.

Reply:
Agreed. It would be useful to convert this into an issue that we tackle in a later PR.